src/smt/theory_str_noodler/formula.cpp

```diff
@@ -1,7 +1,160 @@
 
 #include "formula.h"
+#include "util.h"
+
 
 namespace smt::noodler {
+    void collect_free_vars_rec(const LenNode& root, std::set<BasicTerm>& free_vars, std::set<BasicTerm>& quantified_vars) {
+        switch (root.type) {
+            case LenFormulaType::TRUE:
+            case LenFormulaType::FALSE:
+                return;
+            case LenFormulaType::LEAF: {
+                if (root.atom_val.is_variable()) {
+                    const BasicTerm& var = root.atom_val;
+                    if (!quantified_vars.contains(var)) {
+                        free_vars.insert(var);
+                    }
+                }
+                return;
+            }
+            case LenFormulaType::NOT:
+                collect_free_vars_rec(root.succ.at(0), free_vars, quantified_vars);
+                return;
+            case LenFormulaType::LEQ:
+            case LenFormulaType::LT:
+            case LenFormulaType::EQ:
+            case LenFormulaType::NEQ:
+                collect_free_vars_rec(root.succ.at(0), free_vars, quantified_vars);
+                collect_free_vars_rec(root.succ.at(1), free_vars, quantified_vars);
+                return;
+            case LenFormulaType::PLUS:
+            case LenFormulaType::MINUS:
+            case LenFormulaType::TIMES:
+            case LenFormulaType::AND:
+            case LenFormulaType::OR: {
+                for (const auto& child : root.succ) {
+                    collect_free_vars_rec(child, free_vars, quantified_vars);
+                }
+                return;
+            }
+            case LenFormulaType::FORALL:
+            case LenFormulaType::EXISTS: {
+                auto children_iterator = root.succ.begin();
+                const LenNode& quantified_var_node = *children_iterator;
+                const BasicTerm quantified_var = quantified_var_node.atom_val;
+                children_iterator++;
+
+                auto [target_bucket, did_emplace_happen] = quantified_vars.emplace(quantified_var);
+                assert (did_emplace_happen);  // We want all quantified vars to be unique, at least for now
+
+                for (; children_iterator != root.succ.end(); children_iterator++) {
+                    collect_free_vars_rec(*children_iterator, free_vars, quantified_vars);
+                }
+
+                quantified_vars.erase(quantified_var);
+
+                return;
+            }
+            default:
+                UNREACHABLE();
+        }
+    }
+
+    LenNodePrecision get_resulting_precision_for_conjunction(const LenNodePrecision& p0, const LenNodePrecision& p1) {
+        // If one of the formulae is an underapproximation, then when we find that the resulting conjunction has no solutions,
+        // then we cannot say that for sure - we might have lost some solutions due to imprecision of one of the conjuncts.
+        if (p0 == LenNodePrecision::UNDERAPPROX || p1 == LenNodePrecision::UNDERAPPROX) return LenNodePrecision::UNDERAPPROX;
```

Comment:
I think this is not correct, if one formula is `underapproximation` and the other is `overapproximation`, then we should throw error, they are not comparable. Try drawing the set of models of precise formulas and then make one an underapproximation (so the set of models is smaller for that formula) and the other an overapproximation (so the set of models is larger) and you can be left with a part of models that would not be in precise conjunction (which would mean overapproximation) but also some models are missing from the conjunction (which would mean an underapproximation.

Reply:
I am confused. Say that we have a precisese set of models $X = \\{a, b, c\\}$. We have one formula $\varphi$ that underapproximates the problem solutions with models $\\{ a \\}$., and one formula $\psi$ that overapproximates the solutions, say its models are $\\{a, b, c, d, e \\}$.

A conjunction $\varphi \land \psi$ has solutions $\\{ a \\}$, so $\varphi \land \psi$ is an underapproximation. Why should we throw an error?

Reply:
This function should be doing something different, no?. It should compute the precision of the conjunction of two formulae, each with their own precision. So let's say the precise models of the first formula are {a,b,c}, and  φ underapproximates it into {a,b}, while the precise models of the second formula are {b,c,d}, but ψ overapproximates it into {a,b,c,d,e}. The precise conjunction should have models {b,c}, while the real conjunction is {a,b}, which is neither over- nor underapproximation of the real conjunction.

Reply:
Ah, okay, sorry. My mistake. I was working under the wrong impression that precision is related to the entire branch.

Reply:
Done, I rewrote the function. @jurajsic could you check the function, please?

Reply:
Looks good